File: programmers/2020 KAKAO BLIND RECRUITMENT/기둥과 보 설치(Level3).py
'''
    Code written on Sep 9, 2020
    출처: 프로그래머스 코딩 테스트 연습, https://programmers.co.kr/learn/challenges
    2020 KAKAO BLIND RECRUITMENT: 기둥과 보 설치(Level3)
    link: https://programmers.co.kr/learn/courses/30/lessons/60061
'''
'''
    개선된 코드(삭제할 때만 반복 시행)
'''
def check_condition(x, y, kind, building):
    if kind:    # beam
        if (x,y-1,0) not in building and (x+1,y-1,0) not in building and \
                ((x-1,y,1) not in building or (x+1,y,1) not in building):
            return False
    else:       # column
        if y and (x-1,y,1) not in building and \
                (x,y,1) not in building and (x,y-1,0) not in building:
            return False
    return True

def solution(n, build_frame):
    building = set()
    for x, y, kind, construction in build_frame:
        if construction:    
            if check_condition(x, y, kind, building):
                building.add((x, y, kind))
        else:
            building.remove((x, y, kind))
            for x_, y_, kind_ in building:
                if not check_condition(x_, y_, kind_, building):
                    building.add((x, y, kind))
                    break
    building = map(list, building)
    return sorted(building)


'''
    첫 번째 코드(추가할 때에는 반복할 필요가 없음. 개선이 필요)
'''
'''
def check_condition(building):
    for x, y, kind in building:
        if kind:    # beam
            if (x,y-1,0) not in building and (x+1,y-1,0) not in building and \
                    ((x-1,y,1) not in building or (x+1,y,1) not in building):
                break
        else:       # column
            if y and (x-1,y,1) not in building and \
                    (x,y,1) not in building and (x,y-1,0) not in building:
                break
    else:
        return True
    return False

def solution(n, build_frame):
    building = set()
    for x, y, kind, construction in build_frame:
        if construction:    
            building.add((x, y, kind))
            if not check_condition(building):
                building.remove((x, y, kind))
        else:
            building.remove((x, y, kind))
            if not check_condition(building):
                building.add((x, y, kind))
    building = map(list, building)
    return sorted(building)
'''

File: programmers/2020 KAKAO BLIND RECRUITMENT/기둥과 보 설치(Level3).java
/*
    Code written on Feb 16, 2021
    출처: 프로그래머스 코딩 테스트 연습, https://programmers.co.kr/learn/challenges
    2020 KAKAO BLIND RECRUITMENT: 기둥과 보 설치(Level3)
    link: https://programmers.co.kr/learn/courses/30/lessons/60061
*/
import java.util.Arrays;
import java.util.HashSet;
import java.util.Iterator;

class Solution {
    public int[][] solution(int n, int[][] build_frame) {
        HashSet<Building> buildings = new HashSet<>();

        for (int[] buildingPlan : build_frame) {
            Building building = new Building(buildingPlan[0], buildingPlan[1], buildingPlan[2]);
            if (buildingPlan[3] == 0) {
                buildings.remove(building);
                if (!canBuild(buildings))
                    buildings.add(building);
            } else {
                buildings.add(building);
                if (!canBuild(buildings))
                    buildings.remove(building);
            }
        }

        return toSortedArray(buildings);
    }

    public boolean canBuild(HashSet<Building> buildings) {
        for (Building building: buildings) {
            Building tempBuilding;
            if (building.type == 0) {
                if (building.y == 0)
                    continue;
                tempBuilding = new Building(building.x, building.y, 1);
                if (buildings.contains(tempBuilding))
                    continue;
                tempBuilding = new Building(building.x - 1, building.y, 1);
                if (buildings.contains(tempBuilding))
                    continue;
                tempBuilding = new Building(building.x, building.y - 1, building.type);
                if (!buildings.contains(tempBuilding))
                    return false;
            } else {
                tempBuilding = new Building(building.x, building.y - 1, 0);
                if (buildings.contains(tempBuilding))
                    continue;
                tempBuilding = new Building(building.x + 1, building.y - 1, 0);
                if (buildings.contains(tempBuilding))
                    continue;;
                tempBuilding = new Building(building.x - 1, building.y, building.type);
                Building tempBuilding2 = new Building(building.x + 1, building.y, building.type);
                if (!(buildings.contains(tempBuilding)) || !(buildings.contains(tempBuilding2)))
                    return false;
            }
        }

        return true;
    }

    public int[][] toSortedArray(HashSet<Building> buildings) {
        Iterator<Building> iterator = buildings.iterator();
        int[][] array = new int[buildings.size()][3];

        for (int i = 0; i < array.length; i++) {
            Building building = iterator.next();
            array[i][0] = building.x;
            array[i][1] = building.y;
            array[i][2] = building.type;
        }

        Arrays.sort(array, (o1, o2) -> {
            if (o1[0] != o2[0])
                return o1[0] - o2[0];
            if (o1[1] != o2[1])
                return o1[1] - o2[1];
            return o1[2] - o2[2];
        });

        return array;
    }
}

class Building {
    int x;
    int y;
    int type;

    public Building(int x, int y, int type) {
        this.x = x;
        this.y = y;
        this.type = type;
    }

    @Override
    public boolean equals(Object o) {
        Building building = (Building) o;
        if (x != building.x)
            return false;
        if (y != building.y)
            return false;
        return type == building.type;
    }

    @Override
    public int hashCode() {
        return 100000000 * type + 10000 * x + y;
    }
}
